SSA/Core/Framework.lean

```diff
@@ -220,282 +590,717 @@ args, since there now are equation lemmas for it.
+  | _, ⟨op, sig_eq, eff_leq, args, regArgs⟩ =>
+     ⟨op, sig_eq, eff_leq, args.map varsMap, regArgs⟩
+
+-- TODO: figure out what this comment means?
+/-- @alexkeizer: We need this to generate defeqs properly -/
+def Com.changeVars : Com Op Γ eff ty →
+    (varsMap : Γ.Hom Γ') →
+    Com Op Γ' eff ty
+  |  .ret e => fun varsMap => .ret (varsMap e)
+  |  .lete e body => fun varsMap => .lete (e.changeVars varsMap)
+      (body.changeVars (fun t v => varsMap.snocMap v))
 
-@[simp]
-theorem Expr.denote_changeVars {Γ Γ' : Ctxt Ty}
+/-! simp-lemmas about `changeVars`-/
+
+@[simp] lemma Expr.denote_changeVars {Γ Γ' : Ctxt Ty}
     (varsMap : Γ.Hom Γ')
-    (e : Expr Op Γ ty)
-    (Γ'v : Valuation Γ') :
-    (e.changeVars varsMap).denote Γ'v =
-    e.denote (fun t v => Γ'v (varsMap v)) := by
+    (e : Expr Op Γ eff ty) :
+    (e.changeVars varsMap).denote =
+    fun Γ'v => e.denote (Γ'v.comap varsMap) := by
+  funext Γ'v
   rcases e with ⟨_, rfl, _⟩
-  simp [Expr.denote, Expr.changeVars, HVector.map_map]
+  simp [Expr.denote, Expr.changeVars, HVector.map_map, Valuation.comap]
+
+@[simp] lemma Com.changeVars_ret (v : Var Γ t) :
+    (Com.ret (Op:=Op) (eff := eff) v).changeVars = fun (map : Γ.Hom Δ) => Com.ret (map v) := by
+  funext map
+  simp [changeVars]
+
+@[simp] lemma Com.changeVars_lete (e : Expr Op Γ eff t) (body : Com Op _ eff u) :
+    (Com.lete e body).changeVars
+    = fun (map : Γ.Hom Δ) => Com.lete (e.changeVars map) (body.changeVars map.snocMap) := by
+  simp [changeVars]
+
+-- TODO: this is implied by simpler simp-lemmas, do we need it?
+@[simp] lemma Com.outContext_changeVars_ret (varsMap : Γ.Hom Γ') (c : Com Op Γ eff ty) :
+  ((Com.ret (Op:=Op) (eff := eff) v).changeVars varsMap).outContext = Γ' := by simp
+
+@[simp] lemma Com.denote_changeVars
+    (varsMap : Γ.Hom Γ') (c : Com Op Γ eff ty) :
+    (c.changeVars varsMap).denote =
+    c.denote ∘ (fun V => V.comap varsMap) := by
+  funext Γ'v
+  simp only [Function.comp_apply]
+  induction c using Com.rec' generalizing Γ'v Γ' with
+  | ret x => simp [Com.denote, Com.changeVars, Valuation.comap, *]
+  | lete _ _ ih =>
+    unfold denote
+    simp only [changeVars_lete]
+    split <;> simp_all
+
+@[simp] lemma Com.denote_changeVars' (varsMap : Γ.Hom Γ') (c : Com Op Γ eff ty) :
+    (c.changeVars varsMap).denote = fun V => c.denote (V.comap varsMap) := by
+  simp; rfl
+
+@[simp] lemma Com.outContext_changeVars_hom {map : Γ.Hom Δ} (map_inv : Δ.Hom Γ) :
+    {c : Com Op Γ eff ty} → Ctxt.Hom (outContext (changeVars c map)) (outContext c)
+  | .ret _        => cast (by simp) map_inv
+  | .lete _ body  => cast (by simp) <|
+      Com.outContext_changeVars_hom (map := map.snocMap) map_inv.snocMap (c := body)
+
+-- @[simp] lemma Com.denoteLets_changeVars (map : Γ.Hom Δ) (c : Com Op Γ .pure ty) :
+--     (c.changeVars map).denoteLets = fun (V : Valuation Δ) =>
+--       (c.denoteLets (V.comap map)).comap (fun t' v' => v'.castCtxt (by simp)) := by
+--   sorry
```

Comment:
stray code? or something that needs to be finished later? (can we either uncomment or delete it?)

Reply:
IIRC, this was a proof strategy that we did not pursue. Removed for the sake of cleanliness.